# Data Loading & Cleaning

In this notebook, I'll be Loading in the kaggle dataset "Football Data from Transfermarkt" by David Cariboo. After Loading in the dataset, I'll be merging it's numerous CSV files and combining it into a single dataframe and cleaning that dataframe so that I can begin to perform Exploratory Data Analysis on it.

#### Package Import

Importing the most basic packages that I'll likely be using in this notebook.

In [1]:
import pandas as pd
import numpy as np

## Data Loading

I'll now proceed to load in all my Downloaded CSV files, starting with the player_valuations csv file which contains information about the player valuations at different intervals of time.

### Valuations Dataframe

In [2]:
#Loading the valuations csv;
valuations = pd.read_csv("data1/player_valuations.csv")
#look at a sample of the dataframe
valuations.sample(3)

player_id  last_season             datetime        date    dateweek  \
39531        7346         2014  2010-06-22 00:00:00  2010-06-22  2010-06-21   
86174        7797         2017  2013-01-14 00:00:00  2013-01-14  2013-01-14   
343373     186172         2023  2021-03-10 00:00:00  2021-03-10  2021-03-08   

        market_value_in_eur  n  current_club_id  \
39531               2500000  1              114   
86174                300000  1              410   
343373              3000000  1              589   

       player_club_domestic_competition_id  
39531                                  TR1  
86174                                  IT1  
343373                                 TR1

In [3]:
#look at information on the dataframe
valuations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 440663 entries, 0 to 440662
Data columns (total 9 columns):
 #   Column                               Non-Null Count   Dtype 
---  ------                               --------------   ----- 
 0   player_id                            440663 non-null  int64 
 1   last_season                          440663 non-null  int64 
 2   datetime                             440663 non-null  object
 3   date                                 440663 non-null  object
 4   dateweek                             440663 non-null  object
 5   market_value_in_eur                  440663 non-null  int64 
 6   n                                    440663 non-null  int64 
 7   current_club_id                      440663 non-null  int64 
 8   player_club_domestic_competition_id  440663 non-null  object
dtypes: int64(5), object(4)
memory usage: 30.3+ MB


The dataframe contains the players identification, the time of valuation, his club information, his actual market value in Euros and an ambiguous 'n' collumn. There are no null values and around 440,000 datapoints.

In [4]:
#describe dataframe
valuations.describe()

player_id    last_season  market_value_in_eur         n  \
count  4.406630e+05  440663.000000         4.406630e+05  440663.0   
mean   1.964113e+05    2018.762887         2.357557e+06       1.0   
std    1.793622e+05       3.624305         6.603356e+06       0.0   
min    1.000000e+01    2012.000000         1.000000e+04       1.0   
25%    5.532200e+04    2016.000000         2.000000e+05       1.0   
50%    1.407480e+05    2019.000000         5.000000e+05       1.0   
75%    2.896450e+05    2022.000000         1.600000e+06       1.0   
max    1.166093e+06    2023.000000         2.000000e+08       1.0   

       current_club_id  
count    440663.000000  
mean       4041.891491  
std        9508.375247  
min           3.000000  
25%         368.000000  
50%        1010.000000  
75%        2944.000000  
max       83678.000000

the .describe table tells us that the market values of football players in the dataset varies widely, with an average of approximately 2.36 million EUR. The values range from 10,000 EUR to a maximum of 200 million EUR across multiple seasons.

### Players Dataframe

The players CSV file should give us data on each individual player which will be essential in predicting our target variable `market_value_in_eur`.

In [5]:
#Loading the valuations csv;
players = pd.read_csv("data1/players.csv")
#look at a sample of the dataframe
players.sample(3)

player_id first_name last_name             name  last_season  \
19626     607225    Florent   Sanchez  Florent Sanchez         2023   
21547     700933      Derik   Lacerda    Derik Lacerda         2021   
27029     534376      Marco    Moreno     Marco Moreno         2022   

       current_club_id      player_code country_of_birth  \
19626            54189  florent-sanchez           France   
21547              979    derik-lacerda           Brazil   
27029               13     marco-moreno            Spain   

                    city_of_birth country_of_citizenship  ...   foot  \
19626              Vaulx-en-Velin                 France  ...  right   
21547              Rio de Janeiro                 Brazil  ...   left   
27029  Las Palmas de Gran Canaria                  Spain  ...  right   

      height_in_cm market_value_in_eur highest_market_value_in_eur  \
19626        178.0            700000.0                    700000.0   
21547        191.0            500000.0                    500000.0   
27029        191.0            200000.0                    200000.0   

       contract_expiration_date            agent_name  \
19626       2024-06-30 00:00:00       USFA Management   
21547       2025-06-30 00:00:00                    OF   
27029       2025-06-30 00:00:00  NLN SPORT MANAGEMENT   

                                               image_url  \
19626  https://img.a.transfermarkt.technology/portrai...   
21547  https://img.a.transfermarkt.technology/portrai...   
27029  https://img.a.transfermarkt.technology/portrai...   

                                                     url  \
19626  https://www.transfermarkt.co.uk/florent-sanche...   
21547  https://www.transfermarkt.co.uk/derik-lacerda/...   
27029  https://www.transfermarkt.co.uk/marco-moreno/p...   

      current_club_domestic_competition_id   current_club_name  
19626                                  BE1       RWD Molenbeek  
21547                                  PO1       Moreirense FC  
27029                                  ES1  Atlético de Madrid  

[3 rows x 23 columns]

In [6]:
#.info()
players.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30298 entries, 0 to 30297
Data columns (total 23 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   player_id                             30298 non-null  int64  
 1   first_name                            28334 non-null  object 
 2   last_name                             30298 non-null  object 
 3   name                                  30298 non-null  object 
 4   last_season                           30298 non-null  int64  
 5   current_club_id                       30298 non-null  int64  
 6   player_code                           30298 non-null  object 
 7   country_of_birth                      27607 non-null  object 
 8   city_of_birth                         28093 non-null  object 
 9   country_of_citizenship                29754 non-null  object 
 10  date_of_birth                         30251 non-null  object 
 11  sub_position   

- This dataframe provides us with quite a lot of information on the player like `name`, `date of birth`, `height`, `position` etc aswell as a lot of overlapping information with the valuations dataframe. It also gives us an interesting collumn called `highest_market_value_in_eur` which provides us with the highest the players market value has ever been.
- unlike in player valuations, this dataset has a lot of null values in a few places. We will have to deal with these somewhere down the line aswell as duplicates.

In [7]:
#.describe()
players.describe()

player_id   last_season  current_club_id  height_in_cm  \
count  3.029800e+04  30298.000000     30298.000000  28188.000000   
mean   3.112298e+05   2018.768335      4365.629547    182.233113   
std    2.502184e+05      3.654054     10056.593385      6.834130   
min    1.000000e+01   2012.000000         3.000000     18.000000   
25%    9.526825e+04   2016.000000       403.000000    178.000000   
50%    2.578240e+05   2019.000000      1071.000000    182.000000   
75%    4.655408e+05   2022.000000      3008.000000    187.000000   
max    1.186012e+06   2023.000000     83678.000000    207.000000   

       market_value_in_eur  highest_market_value_in_eur  
count         1.935800e+04                 2.895500e+04  
mean          2.180419e+06                 3.523106e+06  
std           7.096501e+06                 9.217968e+06  
min           1.000000e+04                 1.000000e+04  
25%           1.750000e+05                 2.500000e+05  
50%           3.500000e+05                 7.500000e+05  
75%           1.000000e+06                 2.600000e+06  
max           1.800000e+08                 2.000000e+08

The .describe() method on the players dataframe outputs a similar set of information as from the valuations dataframe. It seems that for each player in the dataframe, the collumn `market_value_in_eur` gives us their most recent valuation rather than a different one at any given point in time.

### Appearances Dataframe

The Appearances should give us details of the appearance of each player in a game.

In [8]:
#Loading the valuations csv;
appearances = pd.read_csv("data1/appearances.csv")
#look at a sample of the dataframe
appearances.sample(3)

appearance_id  game_id  player_id  player_club_id  \
370380   2518662_21924  2518662      21924             128   
788731   3059298_77757  3059298      77757             601   
876637  3069907_355566  3069907     355566           19789   

        player_current_club_id        date      player_name competition_id  \
370380                     128  2015-04-05   Antonio Ghomsi            GR1   
788731                    1031  2018-07-28  Thomas Kaminski            BE1   
876637                     126  2019-02-22      Baris Alici            TR1   

        yellow_cards  red_cards  goals  assists  minutes_played  
370380             0          0      0        0              45  
788731             0          0      0        0              90  
876637             0          0      0        0              23

In [9]:
#.info() to get more information on the CSV file
appearances.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1480937 entries, 0 to 1480936
Data columns (total 13 columns):
 #   Column                  Non-Null Count    Dtype 
---  ------                  --------------    ----- 
 0   appearance_id           1480937 non-null  object
 1   game_id                 1480937 non-null  int64 
 2   player_id               1480937 non-null  int64 
 3   player_club_id          1480937 non-null  int64 
 4   player_current_club_id  1480937 non-null  int64 
 5   date                    1480937 non-null  object
 6   player_name             1480613 non-null  object
 7   competition_id          1480937 non-null  object
 8   yellow_cards            1480937 non-null  int64 
 9   red_cards               1480937 non-null  int64 
 10  goals                   1480937 non-null  int64 
 11  assists                 1480937 non-null  int64 
 12  minutes_played          1480937 non-null  int64 
dtypes: int64(9), object(4)
memory usage: 146.9+ MB


- This dataframe has 1.48 million datapoints indicating 1.48 millions appearances of players at our disposal.
- It contains information on some key player metrics and the team they were playing for at the same aswell as the competition the game was in.
- There are no null values which is a positive.

In [10]:
#describe
appearances.describe()

game_id     player_id  player_club_id  player_current_club_id  \
count  1.480937e+06  1.480937e+06    1.480937e+06            1.480937e+06   
mean   2.994878e+06  1.762232e+05    2.853359e+03            3.621818e+03   
std    5.416510e+05  1.616165e+05    7.007412e+03            9.258317e+03   
min    2.211607e+06  1.000000e+01    1.000000e+00           -1.000000e+00   
25%    2.512765e+06  5.289500e+04    2.890000e+02            3.360000e+02   
50%    2.899821e+06  1.262810e+05    8.550000e+02            9.310000e+02   
75%    3.432247e+06  2.565670e+05    2.441000e+03            2.687000e+03   
max    4.192928e+06  1.166093e+06    8.367800e+04            8.367800e+04   

       yellow_cards     red_cards         goals       assists  minutes_played  
count  1.480937e+06  1.480937e+06  1.480937e+06  1.480937e+06    1.480937e+06  
mean   1.491968e-01  3.840812e-03  9.640856e-02  7.435833e-02    6.973785e+01  
std    3.677885e-01  6.185517e-02  3.319711e-01  2.835161e-01    2.972540e+01  
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00    1.000000e+00  
25%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00    5.300000e+01  
50%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00    9.000000e+01  
75%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00    9.000000e+01  
max    2.000000e+00  1.000000e+00  6.000000e+00  6.000000e+00    1.200000e+02

Ignoring the player information statistics, The dataset provides information on player performance in football matches, including stats such as cards, goals, assists, and minutes played. On average, players receive 0.14 yellow cards & 0.004 red cards aswell as 0.096 goals, and 0.074 assists per game. The average minutes played per game is around 70 minutes. Overall, the stats make sense.

I"ve now loaded in all the csv files, I will need for the time being.

## Data Cleaning

#### Valuations & Players Dataset 

In [11]:
valuations.isna().sum()

player_id                              0
last_season                            0
datetime                               0
date                                   0
dateweek                               0
market_value_in_eur                    0
n                                      0
current_club_id                        0
player_club_domestic_competition_id    0
dtype: int64

No null values here.

In [12]:
players.isna().sum()

player_id                                   0
first_name                               1964
last_name                                   0
name                                        0
last_season                                 0
current_club_id                             0
player_code                                 0
country_of_birth                         2691
city_of_birth                            2205
country_of_citizenship                    544
date_of_birth                              47
sub_position                              173
position                                    0
foot                                     2397
height_in_cm                             2110
market_value_in_eur                     10940
highest_market_value_in_eur              1343
contract_expiration_date                11478
agent_name                              15362
image_url                                   0
url                                         0
current_club_domestic_competition_

Quite a few null values here, they'll have to be dealt with sooner or later.

In [13]:
valuations["player_id"].nunique()

28794

.nunique will return me the number of unique counts of player_ids I have in my valuations dataset. Essentially suggesting that I have valuations for 28,700 individual players with each player having numerous valuations. I"ll have to decide later on whether I want one valuation for each individual player or to keep the numerous valuations.

In [14]:
valuations.isna().sum()

player_id                              0
last_season                            0
datetime                               0
date                                   0
dateweek                               0
market_value_in_eur                    0
n                                      0
current_club_id                        0
player_club_domestic_competition_id    0
dtype: int64

In [15]:
valuations.sample(10)

player_id  last_season             datetime        date    dateweek  \
304854     161549         2020  2020-04-08 00:00:00  2020-04-08  2020-04-06   
266745      60444         2023  2019-06-05 00:00:00  2019-06-05  2019-06-03   
187714      35509         2017  2017-01-09 00:00:00  2017-01-09  2017-01-09   
80529       38414         2022  2012-10-16 00:00:00  2012-10-16  2012-10-15   
276693     334816         2023  2019-06-28 00:00:00  2019-06-28  2019-06-24   
113451     263065         2014  2014-05-08 00:00:00  2014-05-08  2014-05-05   
76838      148369         2013  2012-07-18 00:00:00  2012-07-18  2012-07-16   
223199      32622         2022  2018-01-22 00:00:00  2018-01-22  2018-01-22   
303146      93827         2020  2020-04-08 00:00:00  2020-04-08  2020-04-06   
156842     113612         2013  2016-01-14 00:00:00  2016-01-14  2016-01-11   

        market_value_in_eur  n  current_club_id  \
304854               800000  1             1426   
266745             70000000  1               31   
187714               300000  1             1506   
80529               4000000  1            10484   
276693               250000  1             2553   
113451                50000  1              130   
76838                100000  1              865   
223199              1000000  1              924   
303146              1600000  1               33   
156842               200000  1              416   

       player_club_domestic_competition_id  
304854                                 NL1  
266745                                 GB1  
187714                                 TR1  
80529                                  TR1  
276693                                 SC1  
113451                                 IT1  
76838                                  DK1  
223199                                 TR1  
303146                                  L1  
156842                                 IT1

Get a lengthy sample of 25 of the dataset to get a gauge of how it looks.

## Merging DataFrames

Now that I've got my individual DataFrames which all contain important information that I want in my model, I've got merge them into one singular DataFrame. We can start with the players and valuations DataFrames.

In [16]:
#merge players and valuations dataset
fdf = pd.merge(players, valuations, on="player_id")

Here I'm merging the valuations and player dataframes into one wider dataframe with both the data from both collumns joined on the `player_id` collumn. I"ll continue to concat/merge with the other dataframes until I have my desired dataset for my model.

In [17]:
#do .info on dataframe to get overview
fdf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 440584 entries, 0 to 440583
Data columns (total 31 columns):
 #   Column                                Non-Null Count   Dtype  
---  ------                                --------------   -----  
 0   player_id                             440584 non-null  int64  
 1   first_name                            410797 non-null  object 
 2   last_name                             440584 non-null  object 
 3   name                                  440584 non-null  object 
 4   last_season_x                         440584 non-null  int64  
 5   current_club_id_x                     440584 non-null  int64  
 6   player_code                           440584 non-null  object 
 7   country_of_birth                      423624 non-null  object 
 8   city_of_birth                         430453 non-null  object 
 9   country_of_citizenship                433617 non-null  object 
 10  date_of_birth                         440061 non-null  object 
 11  

Everything seems to be correctly merged on `player_id`. I've got a lot of duplicated collumn which I will have to drop from my DataFrame. This step will be taken after, but to first see if everything was merged correctly we can take a look at Cristiano Ronaldos Data across the DataFrame.

In [18]:
#have a look at ronaldos data in the dataset
fdf[fdf["name"] == "Cristiano Ronaldo"].sample(5)

player_id first_name          last_name               name  \
49617       8198        NaN  Cristiano Ronaldo  Cristiano Ronaldo   
49621       8198        NaN  Cristiano Ronaldo  Cristiano Ronaldo   
49590       8198        NaN  Cristiano Ronaldo  Cristiano Ronaldo   
49601       8198        NaN  Cristiano Ronaldo  Cristiano Ronaldo   
49625       8198        NaN  Cristiano Ronaldo  Cristiano Ronaldo   

       last_season_x  current_club_id_x        player_code country_of_birth  \
49617           2022                985  cristiano-ronaldo         Portugal   
49621           2022                985  cristiano-ronaldo         Portugal   
49590           2022                985  cristiano-ronaldo         Portugal   
49601           2022                985  cristiano-ronaldo         Portugal   
49625           2022                985  cristiano-ronaldo         Portugal   

      city_of_birth country_of_citizenship  ...  \
49617       Funchal               Portugal  ...   
49621       Funchal               Portugal  ...   
49590       Funchal               Portugal  ...   
49601       Funchal               Portugal  ...   
49625       Funchal               Portugal  ...   

      current_club_domestic_competition_id  current_club_name last_season_y  \
49617                                  GB1  Manchester United          2022   
49621                                  GB1  Manchester United          2022   
49590                                  GB1  Manchester United          2022   
49601                                  GB1  Manchester United          2022   
49625                                  GB1  Manchester United          2022   

                  datetime        date    dateweek  market_value_in_eur_y  n  \
49617  2019-12-12 00:00:00  2019-12-12  2019-12-09               75000000  1   
49621  2021-03-23 00:00:00  2021-03-23  2021-03-22               50000000  1   
49590  2008-07-04 00:00:00  2008-07-04  2008-06-30               60000000  1   
49601  2013-06-12 00:00:00  2013-06-12  2013-06-10              100000000  1   
49625  2022-09-15 00:00:00  2022-09-15  2022-09-12               20000000  1   

      current_club_id_y player_club_domestic_competition_id  
49617               985                                 GB1  
49621               985                                 GB1  
49590               985                                 GB1  
49601               985                                 GB1  
49625               985                                 GB1  

[5 rows x 31 columns]

Everything looks well-merged, we can now move onto condensing our DataFrame down into what we want.

## Data Wrangling

#### Refining my DataFrame

In [19]:
fdf2=fdf.loc[:,["player_id","name","last_season_x","current_club_id_x","country_of_citizenship","date_of_birth","sub_position","position","foot","height_in_cm","market_value_in_eur_y","highest_market_value_in_eur","contract_expiration_date","current_club_name","current_club_domestic_competition_id","date"]]

In [20]:
fdf2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 440584 entries, 0 to 440583
Data columns (total 16 columns):
 #   Column                                Non-Null Count   Dtype  
---  ------                                --------------   -----  
 0   player_id                             440584 non-null  int64  
 1   name                                  440584 non-null  object 
 2   last_season_x                         440584 non-null  int64  
 3   current_club_id_x                     440584 non-null  int64  
 4   country_of_citizenship                433617 non-null  object 
 5   date_of_birth                         440061 non-null  object 
 6   sub_position                          439781 non-null  object 
 7   position                              440584 non-null  object 
 8   foot                                  426723 non-null  object 
 9   height_in_cm                          430044 non-null  float64
 10  market_value_in_eur_y                 440584 non-null  int64  
 11  

Our DataFrame has been condensed into a few key features. We can now work to bring in the in-game player statistics aswell as engineer some new features from what we currently have.

#### Turning dates into Datetime

We need to turn a few of our collumns such as `date`,`date_of_birth` & `contract_expiration_date` into the Datetime data-type. 

In [21]:
fdf2["date_of_birth"] = pd.to_datetime(fdf2["date_of_birth"])
fdf2["date"] = pd.to_datetime(fdf2["date"])
fdf2["contract_expiration_date"] = pd.to_datetime(fdf2["contract_expiration_date"])

In [22]:
fdf2.dtypes

player_id                                        int64
name                                            object
last_season_x                                    int64
current_club_id_x                                int64
country_of_citizenship                          object
date_of_birth                           datetime64[ns]
sub_position                                    object
position                                        object
foot                                            object
height_in_cm                                   float64
market_value_in_eur_y                            int64
highest_market_value_in_eur                    float64
contract_expiration_date                datetime64[ns]
current_club_name                               object
current_club_domestic_competition_id            object
date                                    datetime64[ns]
dtype: object

Our desired collumns have now been converted into datetime.

### Basic Feature Engineering

Instead of having the `date_of_birth` collumn, we can have an age collumn that updates as the years go on rather than the date of birth which doesnt change. We can do this by subtracting the date of birth from the date, turning that value into a number of days and dividing by the number of days in a year. To first do this, we need to drop any nulls values in the DataFrame. This will hopefully make age a better predictor of market value in my model than date of birth.

#### Dropping Nulls

In [23]:
#drop nulls
fdf2 = fdf2.dropna()
fdf2.isna().sum()

player_id                               0
name                                    0
last_season_x                           0
current_club_id_x                       0
country_of_citizenship                  0
date_of_birth                           0
sub_position                            0
position                                0
foot                                    0
height_in_cm                            0
market_value_in_eur_y                   0
highest_market_value_in_eur             0
contract_expiration_date                0
current_club_name                       0
current_club_domestic_competition_id    0
date                                    0
dtype: int64

All nulls have now been dropped.

In [24]:
# calculate age at the time of date
fdf2["age"] = (fdf2["date"] - fdf2["date_of_birth"]).dt.days / 365

#round to nearest year
fdf2["age"] = fdf2["age"].round().astype(int)

#check it worked
fdf2["age"].sample(10)

283912    26
279650    19
377580    20
371476    18
303767    28
216017    26
116253    26
355277    34
36826     31
54136     24
Name: age, dtype: int32

That looks like it has worked. Now that I've derived age the `date of birth` collumn is useless. I'll drop it.

In [25]:
#drop DOB
fdf2=fdf2.drop("date_of_birth",axis=1)

#### Contract Time Remaining

We can also do a similar calculation to get the number of days left in the players contract since that might be a better predictor of a players valuation than the date a contract expires. We could subtract the contract expiration date from the current date and turn that into days.

In [26]:
#calculate days left of contract
current_date = pd.to_datetime("today")
fdf2["contract_days_left"] = (fdf2["contract_expiration_date"] - current_date).dt.days
#check if it worked
fdf2["contract_days_left"].sample(10)

184796    1681
256961    -146
304506     950
420935     950
112826     220
183410      38
198986     220
274017    -146
13264     1315
313228    1681
Name: contract_days_left, dtype: int64

This seems to have worked, with positive values indicating that the contract is yet to expire and negative values indicating that the contract expired n-number of days ago.

## Advanced Feature Engineering

My goal is to modify the appearances DataFrame so that it gives us the statistics of each player by year. So for example, Cristiano Ronaldo's in-game statistics in 2012 and 2018 are different and dependent on their games in those years. We can start by getting a `year` & `month` collumn out of our `date` collumn in our main DataFrame.

In [27]:
#turn month and year into datetime
fdf2["month"]=fdf2["date"].dt.month
fdf2["year"] = fdf2["date"].dt.year


We will also create a `year` collumn for the appearances DataFrame and use that aswell as unique player identifiers to group our data by and aggregate the data we want from the player.

In [28]:
#convert date to datetime
appearances["date"] = pd.to_datetime(appearances["date"])

#get year
appearances["year"] = appearances["date"].dt.year

# group by player identifiers and year and get sum of statistics
yearly_stats = appearances.groupby(["player_id", "player_name", "year"]).agg(
    {"goals": "sum", "assists": "sum", "yellow_cards": "sum", "red_cards": "sum", "minutes_played": "sum", "player_club_id": "first","competition_id":"first"}
).reset_index()



This should give us all the stats of the player for that given year aswell as the club and domestic competition of that club. We can check this on Ronaldo.

In [29]:
#test on ronaldo
yearly_stats[yearly_stats["player_name"]=="Cristiano Ronaldo"]

player_id        player_name  year  goals  assists  yellow_cards  \
2702       8198  Cristiano Ronaldo  2012     23        4             6   
2703       8198  Cristiano Ronaldo  2013     59       17            10   
2704       8198  Cristiano Ronaldo  2014     56       21             7   
2705       8198  Cristiano Ronaldo  2015     54       18             5   
2706       8198  Cristiano Ronaldo  2016     42       14             5   
2707       8198  Cristiano Ronaldo  2017     42        9             5   
2708       8198  Cristiano Ronaldo  2018     43       12             3   
2709       8198  Cristiano Ronaldo  2019     25        5             2   
2710       8198  Cristiano Ronaldo  2020     41        6             3   
2711       8198  Cristiano Ronaldo  2021     34        6            11   
2712       8198  Cristiano Ronaldo  2022     13        2             7   

      red_cards  minutes_played  player_club_id competition_id  
2702          0            2268             418            ES1  
2703          1            4218             418            ES1  
2704          1            4309             418            ES1  
2705          1            4578             418            ES1  
2706          0            3906             418            ES1  
2707          0            4260             418            ES1  
2708          1            3885             418            ES1  
2709          0            3469             506            CIT  
2710          0            3416             506            IT1  
2711          0            4239             506            IT1  
2712          0            2602             985            GB1

This has worked very well, We can see all the stats of Cristiano Ronaldo grouped by the years he was in action for. One slight problem we are facing, however, is that the `competition_id` collumn seems to be showing the domestic cup ID sometimes instead of the domestic league ID. Maybe we can link the `player_club_id` collumn with the club_id collumn of the clubs CSV file which we haven't yet loaded.

### Clubs DataFrame

This dataset should include information about the club that we can use merge onto our dataset hopefully giving us some good features to feed into our model.

In [30]:
#Loading the valuations csv;
clubs = pd.read_csv("data1/clubs.csv")

#look at a sample of the dataframe
clubs.sample(3)

club_id      club_code           name domestic_competition_id  \
5        157       kaa-gent       KAA Gent                     BE1   
42      3008      hull-city      Hull City                     GB1   
124      276  hellas-verona  Hellas Verona                     IT1   

     total_market_value  squad_size  average_age  foreigners_number  \
5                   NaN          25         25.0                 13   
42                  NaN          28         25.0                 17   
124                 NaN          33         25.2                 22   

     foreigners_percentage  national_team_players           stadium_name  \
5                     52.0                      8         Ghelamco Arena   
42                    60.7                      6            MKM Stadium   
124                   66.7                      7  Marcantonio Bentegodi   

     stadium_seats net_transfer_record  coach_name  last_season  \
5            19999             +€2.52m         NaN         2023   
42           25586             +€7.50m         NaN         2016   
124          39211            +€43.46m         NaN         2023   

                                                   url  
5    https://www.transfermarkt.co.uk/kaa-gent/start...  
42   https://www.transfermarkt.co.uk/hull-city/star...  
124  https://www.transfermarkt.co.uk/hellas-verona/...

In [31]:
#intial look at the information
clubs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 426 entries, 0 to 425
Data columns (total 16 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   club_id                  426 non-null    int64  
 1   club_code                426 non-null    object 
 2   name                     426 non-null    object 
 3   domestic_competition_id  426 non-null    object 
 4   total_market_value       0 non-null      float64
 5   squad_size               426 non-null    int64  
 6   average_age              388 non-null    float64
 7   foreigners_number        426 non-null    int64  
 8   foreigners_percentage    379 non-null    float64
 9   national_team_players    426 non-null    int64  
 10  stadium_name             426 non-null    object 
 11  stadium_seats            426 non-null    int64  
 12  net_transfer_record      426 non-null    object 
 13  coach_name               0 non-null      float64
 14  last_season              4

From this DataFrame, A few potential good features which have been identified are:

`club_id`

`domestic_competition_id` - the top leagues generally have a lot money at their disposal and in turn have more money to spend on transfers.

`net_transfer_record`- this collumn refers to the the net amount a club has spent or received in player transfers. this could be a good predictor as teams that spend a lot more than they sell are likely to have more valuable players on their books.

`national_team_players` - this collumn refers to the number of players in that club that represent their national team on the international stage. This could be a good predictor as these players will typically be more valuable than a non-national team player.

We can merge this onto our yearly stats collumn.

In [32]:
#merge the yearly stats collumn with a few club stats from the club dataset
merged_data = yearly_stats.merge(clubs[['club_id', 'domestic_competition_id', 'net_transfer_record',"national_team_players"]], left_on='player_club_id', right_on='club_id', how='left')

merged_data[merged_data["player_name"]=="Cristiano Ronaldo"].sample(5)

player_id        player_name  year  goals  assists  yellow_cards  \
2707       8198  Cristiano Ronaldo  2017     42        9             5   
2711       8198  Cristiano Ronaldo  2021     34        6            11   
2703       8198  Cristiano Ronaldo  2013     59       17            10   
2706       8198  Cristiano Ronaldo  2016     42       14             5   
2712       8198  Cristiano Ronaldo  2022     13        2             7   

      red_cards  minutes_played  player_club_id competition_id  club_id  \
2707          0            4260             418            ES1    418.0   
2711          0            4239             506            IT1    506.0   
2703          1            4218             418            ES1    418.0   
2706          0            3906             418            ES1    418.0   
2712          0            2602             985            GB1    985.0   

     domestic_competition_id net_transfer_record  national_team_players  
2707                     ES1           €-122.50m                   19.0  
2711                     IT1            €-21.60m                   15.0  
2703                     ES1           €-122.50m                   19.0  
2706                     ES1           €-122.50m                   19.0  
2712                     GB1           €-151.36m                   22.0

We can now drop the original competition_id collumn that we had before which wasn't correctly displaying the domestic leagues aswell as one of the club_id collumns as we dont need both.

In [33]:
#drop competition and club id
merged_data=merged_data.drop(["competition_id","club_id"],axis=1)

Everything looks good here, it seems to update based on the year which is perfect in the context of our future models.

We can now do our final merge by merging our two dataframes to get our one Dataset.

In [56]:
#set year to integer
merged_data["year"] = merged_data["year"].astype(int)

#merge the dataframes
fdf3 = pd.merge(fdf2, merged_data, how="left", on=["player_id", "year"])

#test on ronaldo
fdf3[fdf3["player_name"]=="Cristiano Ronaldo"].sample(5)


player_id               name  last_season_x  current_club_id_x  \
28276       8198  Cristiano Ronaldo           2022                985   
28269       8198  Cristiano Ronaldo           2022                985   
28256       8198  Cristiano Ronaldo           2022                985   
28279       8198  Cristiano Ronaldo           2022                985   
28267       8198  Cristiano Ronaldo           2022                985   

      country_of_citizenship    sub_position position   foot  height_in_cm  \
28276               Portugal  Centre-Forward   Attack  right         187.0   
28269               Portugal  Centre-Forward   Attack  right         187.0   
28256               Portugal  Centre-Forward   Attack  right         187.0   
28279               Portugal  Centre-Forward   Attack  right         187.0   
28267               Portugal  Centre-Forward   Attack  right         187.0   

       market_value_in_eur_y  ...        player_name goals assists  \
28276               50000000  ...  Cristiano Ronaldo  34.0     6.0   
28269              100000000  ...  Cristiano Ronaldo  43.0    12.0   
28256              100000000  ...  Cristiano Ronaldo  59.0    17.0   
28279               30000000  ...  Cristiano Ronaldo  13.0     2.0   
28267              100000000  ...  Cristiano Ronaldo  42.0     9.0   

      yellow_cards red_cards  minutes_played  player_club_id  \
28276         11.0       0.0          4239.0           506.0   
28269          3.0       1.0          3885.0           418.0   
28256         10.0       1.0          4218.0           418.0   
28279          7.0       0.0          2602.0           985.0   
28267          5.0       0.0          4260.0           418.0   

       domestic_competition_id  net_transfer_record national_team_players  
28276                      IT1             €-21.60m                  15.0  
28269                      ES1            €-122.50m                  19.0  
28256                      ES1            €-122.50m                  19.0  
28279                      GB1            €-151.36m                  22.0  
28267                      ES1            €-122.50m                  19.0  

[5 rows x 29 columns]

### Final Dataframe Cleaning

We now need to clean the data by dropping null values and duplicate values.

In [35]:
#drop nulls
fdf3 = fdf3.dropna()

#drop duplicates
fdf3= fdf3.drop_duplicates()

In [36]:
fdf3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 123594 entries, 10 to 262412
Data columns (total 29 columns):
 #   Column                                Non-Null Count   Dtype         
---  ------                                --------------   -----         
 0   player_id                             123594 non-null  int64         
 1   name                                  123594 non-null  object        
 2   last_season_x                         123594 non-null  int64         
 3   current_club_id_x                     123594 non-null  int64         
 4   country_of_citizenship                123594 non-null  object        
 5   sub_position                          123594 non-null  object        
 6   position                              123594 non-null  object        
 7   foot                                  123594 non-null  object        
 8   height_in_cm                          123594 non-null  float64       
 9   market_value_in_eur_y                 123594 non-null  int

Our DataFrame is now cleaned, but it looks like we have a few wrong datatypes, particularly from the yearly statistics DataFrame we created. They should be integers but are appearing as floats.

In [37]:
#change all the datatypes
fdf3["goals"] = fdf3["goals"].astype(int)
fdf3["assists"] = fdf3["assists"].astype(int)
fdf3["yellow_cards"] = fdf3["yellow_cards"].astype(int)
fdf3["red_cards"] = fdf3["red_cards"].astype(int)
fdf3["minutes_played"]=fdf3["minutes_played"].astype(int)
fdf3["player_club_id"]=fdf3["player_club_id"].astype(int)
fdf3["national_team_players"]=fdf3["national_team_players"].astype(int)


We can anonymize everything by removing the name collumn.

In [38]:
#anonymize everything by removing name
fdf3.drop(["name"],axis=1)

player_id  last_season_x  current_club_id_x country_of_citizenship  \
10           4042           2017                234              Australia   
11           4042           2017                234              Australia   
12           4042           2017                234              Australia   
13           4042           2017                234              Australia   
14           4042           2017                234              Australia   
...           ...            ...                ...                    ...   
262395     371851           2023                410               Slovenia   
262405     586756           2023                410                Ireland   
262406     586756           2023                410                Ireland   
262407     586756           2023                410                Ireland   
262412     925584           2023                410                 France   

       sub_position    position   foot  height_in_cm  market_value_in_eur_y  \
10       Goalkeeper  Goalkeeper   left         194.0                1500000   
11       Goalkeeper  Goalkeeper   left         194.0                1500000   
12       Goalkeeper  Goalkeeper   left         194.0                1500000   
13       Goalkeeper  Goalkeeper   left         194.0                1500000   
14       Goalkeeper  Goalkeeper   left         194.0                1500000   
...             ...         ...    ...           ...                    ...   
262395  Centre-Back    Defender  right         190.0               10000000   
262405   Right-Back    Defender  right         180.0                3500000   
262406   Right-Back    Defender  right         180.0                3500000   
262407   Right-Back    Defender  right         180.0                3500000   
262412  Centre-Back    Defender   left         184.0                 250000   

        highest_market_value_in_eur  ...    player_name goals assists  \
10                        2000000.0  ...     Brad Jones     0       0   
11                        2000000.0  ...     Brad Jones     0       0   
12                        2000000.0  ...     Brad Jones     0       0   
13                        2000000.0  ...     Brad Jones     0       0   
14                        2000000.0  ...     Brad Jones     0       0   
...                             ...  ...            ...   ...     ...   
262395                   10000000.0  ...     Jaka Bijol     1       2   
262405                    3500000.0  ...  Festy Ebosele     0       0   
262406                    3500000.0  ...  Festy Ebosele     0       0   
262407                    3500000.0  ...  Festy Ebosele     0       1   
262412                     300000.0  ...  Axel Guessand     0       0   

       yellow_cards  red_cards  minutes_played  player_club_id  \
10                1          0             720              31   
11                1          0             720              31   
12                0          0             450              31   
13                0          0             450              31   
14                0          0             466              31   
...             ...        ...             ...             ...   
262395            6          0            2511             410   
262405            1          0              53             410   
262406            1          0              53             410   
262407            3          0             870             410   
262412            0          0             149             410   

        domestic_competition_id net_transfer_record  national_team_players  
10                          GB1           €-111.30m                     16  
11                          GB1           €-111.30m                     16  
12                          GB1           €-111.30m                     16  
13                          GB1           €-111.30m                     16  
14                          GB1           €-111.30m                

We should also rename a few collumns which aren't well named.

In [39]:
#rename collumns
fdf3 = fdf3.rename(columns={"current_club_id_x": "most_recent_club_id","market_value_in_eur_y": "market_value","highest_market_value_in_eur":"highest_ever_market_value","domestic_competition_id":"league_id"})

In [40]:
#reset the index
fdf3.reset_index(drop=True)


player_id           name  last_season_x  most_recent_club_id  \
0            4042     Brad Jones           2017                  234   
1            4042     Brad Jones           2017                  234   
2            4042     Brad Jones           2017                  234   
3            4042     Brad Jones           2017                  234   
4            4042     Brad Jones           2017                  234   
...           ...            ...            ...                  ...   
123589     371851     Jaka Bijol           2023                  410   
123590     586756  Festy Ebosele           2023                  410   
123591     586756  Festy Ebosele           2023                  410   
123592     586756  Festy Ebosele           2023                  410   
123593     925584  Axel Guessand           2023                  410   

       country_of_citizenship sub_position    position   foot  height_in_cm  \
0                   Australia   Goalkeeper  Goalkeeper   left         194.0   
1                   Australia   Goalkeeper  Goalkeeper   left         194.0   
2                   Australia   Goalkeeper  Goalkeeper   left         194.0   
3                   Australia   Goalkeeper  Goalkeeper   left         194.0   
4                   Australia   Goalkeeper  Goalkeeper   left         194.0   
...                       ...          ...         ...    ...           ...   
123589               Slovenia  Centre-Back    Defender  right         190.0   
123590                Ireland   Right-Back    Defender  right         180.0   
123591                Ireland   Right-Back    Defender  right         180.0   
123592                Ireland   Right-Back    Defender  right         180.0   
123593                 France  Centre-Back    Defender   left         184.0   

        market_value  ...    player_name goals assists yellow_cards red_cards  \
0            1500000  ...     Brad Jones     0       0            1         0   
1            1500000  ...     Brad Jones     0       0            1         0   
2            1500000  ...     Brad Jones     0       0            0         0   
3            1500000  ...     Brad Jones     0       0            0         0   
4            1500000  ...     Brad Jones     0       0            0         0   
...              ...  ...            ...   ...     ...          ...       ...   
123589      10000000  ...     Jaka Bijol     1       2            6         0   
123590       3500000  ...  Festy Ebosele     0       0            1         0   
123591       3500000  ...  Festy Ebosele     0       0            1         0   
123592       3500000  ...  Festy Ebosele     0       1            3         0   
123593        250000  ...  Axel Guessand     0       0            0         0   

        minutes_played  player_club_id  league_id  net_transfer_record  \
0                  720              31        GB1            €-111.30m   
1                  720              31        GB1            €-111.30m   
2                  450              31        GB1            €-111.30m   
3                  450              31        GB1            €-111.30m   
4                  466              31        GB1            €-111.30m   
...                ...             ...        ...                  ...   
123589            2511             410        IT1              +€8.56m   
123590              53             410        IT1              +€8.56m   
123591              53             410        IT1              +€8.56m   
123592             870             410        IT1              +€8.56m   
123593             149             410        IT1              +€8.56m   

       national_team_players  
0                         16  
1                         16  
2                         16  
3                         16  
4                         16  
...                      ...  
123589                     8  
123590                     8  
123591                     8  
123592                

In [41]:
#Drop useless collumns
fdf3=fdf3.drop(["contract_expiration_date","current_club_domestic_competition_id","current_club_name","player_name","name"],axis=1)

Ive dropped these columns as I"ve deemed them useless for my further analysis as I"ve got collumns that are either a better representation of the data or they just dont serve any purpose.

#### Fixing `Net_Transfer_Record` formatting

For the sake of our future model, we need to turn the collumn `net_transfer_record` into its float counterpart rather than the string/object datatype that it's in right now.

In [42]:
fdf3["net_transfer_record"].sample(5)

123826         +-0
62299          +-0
242026    +€23.10m
115658    €-38.70m
103561     +€8.56m
Name: net_transfer_record, dtype: object

I'll first have to remove the value units symbol which should be "€" for all non zero values.

In [43]:
#check how many values contain the euro symbol
print(fdf3["net_transfer_record"].str.contains("€").sum())

#check pound sign too
print(fdf3["net_transfer_record"].str.contains("£").sum())


105581
0


In [44]:
#check how many zero values i have
fdf3[fdf3["net_transfer_record"]=="+-0"].shape

(18013, 24)

In [45]:
fdf3.shape

(123594, 24)

105581 + 18013 adds up to 123594 which matches up perfectly with my row number which is good news as it means that all values are either 0 or set to a standardized euro value.

In [46]:
#remove euro symbol from my values
fdf3["net_transfer_record"]=fdf3["net_transfer_record"].str.replace("€","")

In [47]:
fdf3["net_transfer_record"].sample(20)

223817     -13.11m
10569        -350k
182271      +2.40m
152291      -6.50m
244637       -300k
228833    -146.50m
206488      -7.94m
165596       +200k
128333      -2.30m
246154       -500k
60887       -2.60m
145490      -8.60m
215102       +100k
101582      -5.00m
109209         +-0
42473       +6.70m
77801     -122.50m
182615       +510k
21925          +-0
88912      +45.17m
Name: net_transfer_record, dtype: object

That seems to have worked well.

In [48]:
#display count of where "m" is represented
print(fdf3["net_transfer_record"].str.contains("m").sum())

#display count of where "k" is represented
print(fdf3["net_transfer_record"].str.contains("k").sum())

85991
19590


Again this adds up to 105581 indicating that all non-zero values are represented with an m (millions) or k (thousands).

In [49]:
#replace the k with a 1000x multiplier
fdf3["net_transfer_record"] = fdf3["net_transfer_record"].str.replace('k', 'e3', regex=True)

#replace the m with a 1,000,000x multiplier
fdf3["net_transfer_record"] = fdf3["net_transfer_record"].str.replace('m', 'e6', regex=True)

# Remove the "+-" signs before 0
fdf3["net_transfer_record"] = fdf3["net_transfer_record"].str.replace('[+-](?=0$)', '', regex=True)
fdf3["net_transfer_record"] = pd.to_numeric(fdf3["net_transfer_record"])  # Convert to float

In [50]:
fdf3["net_transfer_record"].head(10)

10   -111300000.0
11   -111300000.0
12   -111300000.0
13   -111300000.0
14   -111300000.0
15   -111300000.0
18      2680000.0
19      2680000.0
20      -500000.0
21      -500000.0
Name: net_transfer_record, dtype: float64

In [51]:
#look at my cleaned dataset on cristiano ronaldo
fdf3[fdf3["player_id"]==8198]

player_id  last_season_x  most_recent_club_id country_of_citizenship  \
28253       8198           2022                  985               Portugal   
28254       8198           2022                  985               Portugal   
28255       8198           2022                  985               Portugal   
28256       8198           2022                  985               Portugal   
28257       8198           2022                  985               Portugal   
28258       8198           2022                  985               Portugal   
28259       8198           2022                  985               Portugal   
28260       8198           2022                  985               Portugal   
28261       8198           2022                  985               Portugal   
28262       8198           2022                  985               Portugal   
28263       8198           2022                  985               Portugal   
28264       8198           2022                  985               Portugal   
28265       8198           2022                  985               Portugal   
28266       8198           2022                  985               Portugal   
28267       8198           2022                  985               Portugal   
28268       8198           2022                  985               Portugal   
28269       8198           2022                  985               Portugal   
28270       8198           2022                  985               Portugal   
28271       8198           2022                  985               Portugal   
28272       8198           2022                  985               Portugal   
28273       8198           2022                  985               Portugal   
28274       8198           2022                  985               Portugal   
28275       8198           2022                  985               Portugal   
28276       8198           2022                  985               Portugal   
28277       8198           2022                  985               Portugal   
28278       8198           2022                  985               Portugal   
28279       8198           2022                  985               Portugal   
28280       8198           2022                  985               Portugal   
28281       8198           2022                  985               Portugal   

         sub_position position   foot  height_in_cm  market_value  \
28253  Centre-Forward   Attack  right         187.0      90000000   
28254  Centre-Forward   Attack  right         187.0     100000000   
28255  Centre-Forward   Attack  right         187.0     100000000   
28256  Centre-Forward   Attack  right         187.0     100000000   
28257  Centre-Forward   Attack  right         187.0     100000000   
28258  Centre-Forward   Attack  right         187.0     100000000   
28259  Centre-Forward   Attack  right         187.0     120000000   
28260  Centre-Forward   Attack  right         187.0     120000000   
28261  Centre-Forward   Attack  right         187.0     120000000   
28262  Centre-Forward   Attack  right         187.0     110000000   
28263  Centre-Forward   Attack  right         187.0     110000000   
28264  Centre-Forward   Attack  right         187.0     110000000   
28265  Centre-Forward   Attack  right         187.0     110000000   
28266  Centre-Forward   Attack  right         187.0     100000000   
28267  Centre-Forward   Attack  right         187.0     100000000   
28268  Centre-Forward   Attack  right         187.0     120000000   
28269  Centre-Forward   Attack  right         187.0     100000000   
28270  Centre-Forward   Attack  right         187.0     100000000   
28271  Centre-Forward   Attack  right         187.0      90000000   
28272  Centre-Forward   Attack  right         187.0      75000000   
28273  Centre-Forward   Attack  right         187.0      60000000   
28274  Centre-Forward   Attack  right         187.0      60000000   
28275  Centre-Forward   Attack  right         187

### Data Dictionary

1. `player_id` (int64): A unique identifier for each football player.

2. `last_season_x` (int64): The last season the player participated in.

3. `country_of_citizenship` (object): The player's country of citizenship.

4. `position` (object): The player's primary playing position (e.g., forward, midfielder, defender).

5. `sub_position` (object): The specific position the player plays within a broader position category.

6. `foot` (object): The preferred foot for playing (e.g., left, right).

7. `height_in_cm` (float64): The player's height in centimeters.

8. `market_value_in_eur` (int64): The market value of the player in euros, which is the target variable for your analysis.

9. `highest_market_value_in_eur` (float64): The player's highest market value in euros.

10. `most_recent_club_id` (object): The ID of the player's most recent club.

11. `date` (datetime64): A date associated with the player's record.

12. `age` (int32): The player's age, calculated from their date of birth.

13. `contract_days_left` (int64): The number of days left on the player's current contract.

14. `year` (int64): The year associated with the player's record.

15. `goals` (int32): The number of goals scored by the player that year.

16. `assists` (int32): The number of assists made by the player that year.

17. `yellow_cards` (int32): The number of yellow cards received by the player that year.

18. `red_cards` (int32): The number of red cards received by the player that year.

19. `minutes_played` (int32): The total number of minutes the player has played that year.

20. `net_transfer_record` (float64): The net transfer record of that specific club.

21. `national_team_players` (int32): The number of national team players for that specific club.

22. `month` (int64): The month associated with the player's record.

23. `player_club_id` (int64): The ID of the player's club.

24. `league_id` (object): The ID of the league.

Another version of this can be found in the readme file in the project repository.


### Data Export

In [52]:
import pickle
fdf3.to_pickle("data1/fdf3.pkl")
